arrow-core-test/src/main/kotlin/arrow/core/test/laws/FxLaws.kt

```diff
@@ -0,0 +1,166 @@
+package arrow.core.test.laws
+
+import arrow.Kind
+import arrow.core.EagerBind
+import arrow.core.test.concurrency.SideEffect
+import arrow.core.test.generators.throwable
+import arrow.typeclasses.suspended.BindSyntax
+import io.kotlintest.fail
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+private typealias EagerFxBlock<F, A> = (suspend EagerBind<F>.() -> A) -> Kind<F, A>
+private typealias SuspendFxBlock<F, A> = suspend (suspend BindSyntax<F>.() -> A) -> Kind<F, A>
+
+object FxLaws {
+
+  fun <F, A> laws(
+    pureGen: Gen<Kind<F, A>>, // TODO cannot specify or filter a pure generator, so we need to require an additional one
+    G: Gen<Kind<F, A>>,
+    fxBlock: EagerFxBlock<F, A>,
+    sfxBlock: SuspendFxBlock<F, A>
+  ): List<Law> = listOf(
+    Law("non-suspended fx is lazy") { nonSuspendedIsLazy(G, fxBlock) },
+    Law("suspended fx is lazy") { suspendedIsLazy(G, sfxBlock) },
+    Law("non-suspended fx can bind immediate values") { nonSuspendedCanBindImmediate(G, fxBlock) },
+    Law("non-suspended fx can bind immediate exceptions") { nonSuspendedCanBindImmediateException(pureGen, fxBlock) },
+    Law("suspended fx can bind immediate values") { suspendedCanBindImmediateValues(G, sfxBlock) },
+    Law("suspended fx can bind suspended values") { suspendedCanBindSuspendedValues(G, sfxBlock) },
+    Law("suspended fx can bind immediate exceptions") { suspendedCanBindImmediateExceptions(pureGen, sfxBlock) },
+    Law("suspended fx can bind suspended exceptions") { suspendedCanBindSuspendedExceptions(pureGen, sfxBlock) }
+  )
+
+  private suspend fun <F, A> nonSuspendedIsLazy(G: Gen<Kind<F, A>>, fxBlock: EagerFxBlock<F, A>) {
+    forAll(G) { f: Kind<F, A> ->
+      val effect = SideEffect()
```

Comment:
@nomisRev I'm unsure if this is expected, but all new fx blocks are not lazy by default it seems. Or I'm doing smth wrong :/

Reply:
All the data types except for Eval are eager. Which we can make eager by rewrapping in `Eval`, which is what we do now too I think? 🤔

Reply:
Right, it was only eval which we made lazy + IO 👌 I'll remove this tests then and leave Eval as an special case, although it might be confusing for people who don't know which types can be lazy or not, maybe it would make sense to typealias the syntax for them to LazySyntax (similarly as EagerSyntax now)?. For example I'm thinking on SequenceK here 🤔

Reply:
`fx` is a DSL to avoid nesting with `flatMap`, so the users should still be aware of what they're doing with the actual data types.

`SequenceK` will require `flatMap` rewrite afaik, or a multi-shot continuation solution.